finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -0,0 +1,340 @@
+    /// Lowest curretnly available session.
+    lower_bound: SessionId,
+}
+
+impl<AP, FS> VerifierCache<AP, FS>
+where
+    AP: AuthorityProvider<BlockNumber>,
+    FS: FinalizationStatus,
+{
+    pub fn new(
+        session_period: SessionPeriod,
+        finalization_status: FS,
+        authority_provider: AP,
+        cache_size: SessionId,
+    ) -> Self {
+        Self {
+            sessions: HashMap::new(),
+            session_period,
+            finalization_status,
+            authority_provider,
+            cache_size,
+            lower_bound: SessionId(0),
+        }
+    }
+}
+
+impl<AP, FS> VerifierCache<AP, FS>
+where
+    AP: AuthorityProvider<BlockNumber>,
+    FS: FinalizationStatus,
+{
+    /// Download authorities for session and store in cache. It needs to be first session,
+    /// or first block from previous session needs to be finalized.
+    /// Otherwise nothing is downloaded.
+    fn download(&mut self, session_id: SessionId) {
+        let maybe_session = match session_id {
+            SessionId(0) => self.authority_provider.authority_data(0),
+            SessionId(id) => {
+                let prev_first = first_block_of_session(SessionId(id - 1), self.session_period);
+                self.authority_provider.next_authority_data(prev_first)
+            }
+        };
+
+        if let Some(session) = maybe_session {
+            self.sessions.insert(session_id, session.into());
+        };
+    }
+
+    /// Prune all sessions with number smaller than `session_id`
+    fn prune(&mut self, session_id: SessionId) {
+        self.sessions.retain(|&id, _| id >= session_id);
+        self.lower_bound = session_id;
+    }
+
+    /// Returns session verifier for block number if available. Updates cache if necessary.
+    pub fn get(&mut self, number: BlockNumber) -> Result<&SessionVerifier, CacheError> {
+        let session_id = session_id_from_block_num(number, self.session_period);
+
+        if session_id < self.lower_bound {
+            return Err(CacheError::SessionTooOld(session_id, self.lower_bound));
+        }
+
+        // We are sure about authorities in all session that have first block from previous session finalized.
+        let upper_bound = SessionId(
+            session_id_from_block_num(
+                self.finalization_status.finalized_number(),
+                self.session_period,
+            )
+            .0 + 1,
+        );
+        if session_id > upper_bound {
+            return Err(CacheError::SessionInFuture(session_id, upper_bound));
+        }
+
+        if session_id.0 >= self.cache_size.0 + self.lower_bound.0 {
+            self.prune(SessionId(
+                session_id.0.saturating_sub(self.cache_size.0) + 1,
+            ));
+        }
+
+        if !self.sessions.contains_key(&session_id) {
+            self.download(session_id);
+        }
+
+        self.sessions
+            .get(&session_id)
+            .ok_or(CacheError::UnknownSession)
```

Comment:
this error is strange. The error here really means that we don't know the authorities for this session which is known as we provide the block number to the function itself.

Reply:
Well, to be honest this error should not happen, unless something with `AuthorityProvider` is broken. I can rename it to `UnknownAuthorities` if you think that is better.

Reply:
It is a little bit tricky, to use entry here. It takes a mutable ref and then only if it is not occupied we would want to download. So download can't take the ref of `VerifierCache`... I would leave it as it is. Right now we know, that this error is because of `AuthorityProvider`

Reply:
You can make `download` into a function taking (mut?)ref to `AuthorityProvider` rather than a method, and returning an `Option` (then mapped into a result with the auth provider error). The `Entry` will only take a mutref to the cache itself, so the borrow checker shouldn't complain.

Reply:
Yep, I feel like this complicates the code a little bit, but will add that. If in future (after refactor) authority provider would do all this logic all would be simpler.